src/hotp/hotp.c

```diff
@@ -493,11 +493,13 @@ uint8_t* page = (uint8_t *) current_slot_address;
 
     memcpy (page_buffer, page, SLOT_PAGE_SIZE);
 
-    // check if the secret from the tool is empty and if it is use the old
-    // secret
-uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
-
-    if (secret[0] == 0)
+    // Check if the secret from the tool is empty and if it is use the old secret
+    // Secret could begin with 0x00, so checking the whole secret before keeping the old one in mandatory
+    // The while last loop points secret[-1]. Well, it's right, == result does not matter and mem exits, we need a -1
+    uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
+    int8_t i = 20;
+    while (i >= 0 && secret[--i] == 0x00);
```

Comment:
"memory that doesn't belong to us" was maybe an exaggeration here. I meant that the straightforward semantics of the secret variable is: "An array of 20 uint8_t". But then we access the byte before it.

Unnecessary tricks like that are groundwork for future problems. We should be very obvious about guarantees and requirements.

I'm sure it is fine right now, but changes in seemingly unrelated code and an architecture/toolchain switch may make it wrong in the future without anyone noticing. The data pointer is passed in from the outside, so we shouldn't assume anything about it. If SECRET_OFFSET becomes 0 in a future version, we are actually accessing memory we really don't know anything about in the scope of this function. And we can't expect someone who changes code somewhere else to remember it.

Early exit doesn't matter, since this takes no time, but it is provided anyway by the "empty && i<20".